packages/drivers/odsp-driver/src/snapshotRepresentation.ts

```diff
@@ -0,0 +1,486 @@
+*/
+const binaryBytesToCodeMap = {
+    0: 32,
+    1: 33,
+    2: 34,
+    4: 35,
+    8: 16,
+};
+
+/**
+ * Calculate how many bytes are required to encode an integer. This is always multiple of 2.
+ * So if 6 bytes are required to store an integer than it will return 8.
+ * @param num - number to encode.
+ */
+function calcLength(numArg: number) {
+    let num = numArg;
+    let lengthLen = 0;
+    while (num > 0) {
+        num = Math.floor(num / 256);
+        lengthLen++;
+    }
+    let res = 0;
+    let index = 0;
+    while (res < lengthLen) {
+        res = Math.pow(2, index);
+        index++;
+    }
+    return res;
+}
+
+export function iteratePairs<T>(it: IterableIterator<T>) {
+    const res: IterableIterator<[T, T]> = {
+        next: () => {
+            const a = it.next();
+            if (a.done) {
+                return { value: undefined, done: true };
+            }
+            const b = it.next();
+            assert(b.done !== true, "Should be a pair");
+            return { value: [a.value, b.value], done: b.done };
+        },
+        [Symbol.iterator]: () => { return res; },
+    };
+    return res;
+}
+
+/**
+ * Helper function that returns iterator from an object
+ * @param obj - object that supports iteration
+ */
+export function iterate<T>(obj: {[Symbol.iterator]: () => IterableIterator<T>}) {
+    return obj[Symbol.iterator]();
+}
+
+/**
+ * Base class to represent binary blob element.
+ * Binary blob is one of three types supported as a leaf node of a tree.
+ * Note: concrete implementations (derived classes) are not exposed from this module
+ */
+export abstract class BlobCore {
+    public abstract get buffer(): Uint8Array;
+    /**
+     * Represents a blob.
+     * @param useStringMarkerCode - Represents is the string marker code should be used when representing or
+     * the byte marker code.
+     */
+    constructor(private readonly useStringMarkerCode: boolean = false) {}
+
+    public toString(encoding = stringEncoding) {
+        return Uint8ArrayToString(this.buffer, encoding);
+    }
+
+    public write(buffer: WriteBuffer) {
+        const data = this.buffer;
+        const lengthLen = calcLength(data.length);
+        // Write Marker code.
+        buffer.write(this.useStringMarkerCode ? stringBytesToCodeMap[lengthLen] : binaryBytesToCodeMap[lengthLen]);
+        // Write actual data if length greater than 0, otherwise Marker Code is enough.
+        if (lengthLen > 0) {
+            buffer.write(data.length, lengthLen);
+            for (const element of data) {
+                buffer.write(element);
+            }
+        }
+    }
+}
+
+/**
+ * "deep copy" blob, holds to binary data passed in
+ * It is called deep copy as a counter-part to BlobShallowCopy, which
+ * is a reference to underlying binary stream (ReadBuffer).
+*/
+class BlobDeepCopy extends BlobCore {
+    constructor(protected readonly data: Uint8Array, useStringMarkerCode: boolean = false) {
```

Comment:
Can you please add comment with @param description of each argument?
It's a bit hard to follow when and why useStringMarkerCode should be true or false. Is this developer choice? Is it mentioned somewhere in payload what need to be used in a given response?

Reply:
Yes, we have different codes whether the content that follows is utf8 string format or just the general byte[] representation. Added comments and also changed the name. This does not matter when we read the snapshot. It is just used while writing which tells us whether we want to use encode in utf8 string or not.